### PR TITLE
Add missing data summary to all confirm pages (personal details)

### PR DESCRIPTION
**Pull request by darokel**
### Context

- https://trello.com/c/yxJ5Aw4Z/2366-l-missing-information-ui

### Changes proposed in this pull request

This PR implements an approach around handling missing data for confirm pages (this is for the personal details to begin with). The following changes are proposed:

- Form object implements a `missing_fields` validation context
- The method above can be extracted into a module which is included in every form object. Will return keys corresponding to the missing fields (based on existing validation rules)
- The `MissingDataView` then gets the form instance and iterate over these missing fields, looking up the correct display value used by the summary component render the field name correctly (and anchor to it's row correctly)
- The look up display names are re-referenced from other yaml keys in the translation files so they're all change-able from one place.

Screenshot 2021-08-02 at 18 52 44

### Guidance to review

https://register-pr-1219.london.cloudapps.digital/trainees/NwzJsDTzfHVncTerTGomu7LH/personal-details/confirm

Select Provider A

### CODE REVIEW

**Comment by darokel** [comment]
[[L] Missing information UI (Personal details)](https://trello.com/c/yxJ5Aw4Z/2366-l-missing-information-ui-personal-details)

File: app/forms/personal_details_form.rb [hunk 1]
```diff
@@ -81,6 +81,12 @@ def nationality_ids
       end
   end
 
+  def missing_fields
+    (fields.select { |k, v| k if v.blank? }.keys & FIELDS).tap do |field|
+      field << :date_of_birth unless date_of_birth.is_a?(Date)
+    end
+  end
```
Comment (defong) [inline app/forms/personal_details_form.rb:88]:
```suggestion
  def missing_fields
    if !self.valid?(:missing_data)
      self.errors.attribute_names      
    end  
  end
```

This just looks like another form of validation to me.

**Review by defong — CHANGES_REQUESTED** [review]
This to me looks like just another type of validation context, https://guides.rubyonrails.org/active_record_validations.html#on

Implementing a `custom` method like `missing_fields` means we run the risk of 
https://register-pr-1219.london.cloudapps.digital/trainees/vVJWx9jJSc9mU4uzzrnqVRfv/personal-details/confirm

File: app/forms/personal_details_form.rb (hunk 1 above)
Comment (darokel) [reply app/forms/personal_details_form.rb]: Good shout have updated

File: app/components/personal_details/view.rb [hunk 2]
```diff
@@ -12,37 +12,48 @@ def initialize(data_model:)
       @not_provided_copy = I18n.t("components.confirmation.not_provided")
```
[the remark this answers is not in the document]
Comment (darokel) [reply app/components/personal_details/view.rb]: done

File: app/views/trainees/confirm_details/show.html.erb [hunk 3]
```diff
@@ -4,6 +4,15 @@
...
 
+<%= render InformationSummary::View.new(renderable: @missing_data_view&.missing_data?) do |component| %>
+  <% component.header  do %>
+    <%= @missing_data_view.summary_content %>
+  <% end %>
+
+  <%= @missing_data_view.missing_items_content %>
+<% end %>
+
+
```
Comment (defong) [inline app/views/trainees/confirm_details/show.html.erb:15]:
```suggestion
```

**Review by defong — APPROVED** [review]

**Comment by bbelward** [comment]
looks good, nice work!

**Pull request merged**